From 65f73807ce1d47c344dea4f9334fd91db04b62bb Mon Sep 17 00:00:00 2001
From: Jerry Cao <jerry.cao@amlogic.com>
Date: Tue, 11 Mar 2014 17:02:56 -0700
Subject: [PATCH 3736/5965] inet: Fix ipv6_recv_error callback function
 signature in pingv6_ops

Upstream commit bceaa90240b6019ed73b49965eac7d167610be69 ("inet: prevent
leakage of uninitialized memory to user in recv syscalls") changed
ipv6_recv_error() function signature.

The change is partially patched in 3.10 but the ipv6 portion is not
ported. Backport the function signature change to fix broken build.
---
 include/net/ping.h | 3 ++-
 net/ipv4/ping.c    | 2 +-
 net/ipv6/ping.c    | 3 ++-
 3 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/include/net/ping.h b/include/net/ping.h
index 9242fa090d3d..a494404dbaa4 100644
--- a/include/net/ping.h
+++ b/include/net/ping.h
@@ -31,7 +31,8 @@
 
 /* Compatibility glue so we can support IPv6 when it's compiled as a module */
 struct pingv6_ops {
-	int (*ipv6_recv_error)(struct sock *sk, struct msghdr *msg, int len);
+	int (*ipv6_recv_error)(struct sock *sk, struct msghdr *msg, int len,
+			       int *addr_len);
 	int (*ip6_datagram_recv_ctl)(struct sock *sk, struct msghdr *msg,
 				     struct sk_buff *skb);
 	int (*icmpv6_err_convert)(u8 type, u8 code, int *err);
diff --git a/net/ipv4/ping.c b/net/ipv4/ping.c
index 876f72f39797..eb1fcace7293 100644
--- a/net/ipv4/ping.c
+++ b/net/ipv4/ping.c
@@ -850,7 +850,7 @@ int ping_recvmsg(struct kiocb *iocb, struct sock *sk, struct msghdr *msg,
 			return ip_recv_error(sk, msg, len, addr_len);
 #if IS_ENABLED(CONFIG_IPV6)
 		} else if (family == AF_INET6) {
-			return pingv6_ops.ipv6_recv_error(sk, msg, len);
+			return pingv6_ops.ipv6_recv_error(sk, msg, len, addr_len);
 #endif
 		}
 	}
diff --git a/net/ipv6/ping.c b/net/ipv6/ping.c
index 858788da5a2f..f262789601e6 100644
--- a/net/ipv6/ping.c
+++ b/net/ipv6/ping.c
@@ -57,7 +57,8 @@ static struct inet_protosw pingv6_protosw = {
 
 
 /* Compatibility glue so we can support IPv6 when it's compiled as a module */
-int dummy_ipv6_recv_error(struct sock *sk, struct msghdr *msg, int len)
+int dummy_ipv6_recv_error(struct sock *sk, struct msghdr *msg, int len,
+			  int *addr_len)
 {
 	return -EAFNOSUPPORT;
 }
-- 
2.19.0

